File: js/pig-latin.js
function translate(phrase, translation=[]){
  phrase.split(" ").forEach(function(word){
    translation.push(pigLatinify(word));
  });
  return translation.join(" ");
};

function pigLatinify(word){
  if (word[0].match(/[aeiou]/) && !(word[0]="u" && word[word.length-1]=="q")){
    return word + 'ay';;
  }
  else{
    return pigLatinify(word.substring(1, word.length) + word[0]);
  };
};

module.exports.translate = translate;

File: ruby/pig_latin.rb
class PigLatin

  def self.translate(word_or_phrase)
    word_or_phrase.split.map {|word| self.mod_word(word)}.join(" ")
  end

  def self.mod_word(word, displaced = "")
    displaced += word.slice!(0) until self.end_of_sound?(word, displaced) || word.length == 1
    word + displaced + "ay"
  end

  def self.end_of_sound?(word, displaced)
    ("aeiou".include?(word[0]) && displaced[-1] != "q") || ["yt", "xr"].include?(word[0..1])
  end

end
